misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt

```diff
@@ -0,0 +1,119 @@
+package misk.jdbc
+
+import com.squareup.moshi.Moshi
+import misk.moshi.adapter
+import misk.okio.split
+import net.ttddyy.dsproxy.ExecutionInfo
+import net.ttddyy.dsproxy.QueryInfo
+import net.ttddyy.dsproxy.listener.QueryExecutionListener
+import net.ttddyy.dsproxy.support.ProxyDataSource
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import java.io.EOFException
+import javax.inject.Singleton
+import javax.sql.DataSource
+
+data class Instruction(
+  val Opcode: String,
+  val Input: Instruction?
+) {
+  val isScatter: Boolean
+    get() {
+      return Opcode == "SelectScatter" || Input?.isScatter ?: false
+    }
+}
+
+data class QueryPlan(
+  val Original: String,
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+/**
+ * Throws a [CrossShardQueryException] for scatter queries that doesn't have a lookup vindex.
+ * Note: Current implementation is not thread safe and will not work in production.
+ */
+@Singleton
+class CrossShardQueryDetector(
+  val okHttpClient: OkHttpClient,
+  val moshi: Moshi,
+  val config: DataSourceConfig
+) : DataSourceDecorator {
+
+  private val listener = Listener()
+
+  override fun decorate(dataSource: DataSource): DataSource {
+    if (config.type != DataSourceType.VITESS) return dataSource
```

Comment:
should this be called out in the API? Maybe the classname?

Reply:
I renamed it to "VitessScatterException"